Add defensive check to prevent Mojito gets loading multi times

allmywant opened the pull request.

allmywant committed:
Update nodejs.yml

allmywant committed:
Update dep integrities

dependabot[bot] committed:
Bump minimist from 1.2.5 to 1.2.6

Bumps [minimist](https://github.com/substack/minimist) from 1.2.5 to 1.2.6.
- [Release notes](https://github.com/substack/minimist/releases)
- [Commits](https://github.com/substack/minimist/compare/1.2.5...1.2.6)

---
updated-dependencies:
- dependency-name: minimist
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

allmywant committed:
Bump packages

allmywant committed:
Update package-lock.json

allmywant committed:
Update integrity of glob-parent

allmywant committed:
Merge branch 'master' of https://github.com/allmywant/mojito-js-delivery

allmywant committed:
Mojito library feature requests & updates

Update integrity of decode-uri-component

Downgrade decode-uri-component to 0.3.0

Update integrity of decode-uri-component package

Update integrity of glob-parent package

Update integrity of glob-parent package

Bump the Node packages & resolve any security updates

Bump the GitHub unit test suite to run against later Node.JS versions

allmywant committed:
Bump the Node packages & resolve any security updates

Bump the GitHub unit test suite to run against later Node.JS versions

allmywant committed:
Update integrity of glob-parent package

allmywant committed:
Update integrity of glob-parent package

allmywant committed:
Update integrity of decode-uri-component package

allmywant committed:
Downgrade decode-uri-component to 0.3.0

allmywant committed:
Update integrity of decode-uri-component

allmywant committed:
Mojito library feature requests & updates

allmywant committed:
Merge branch 'master' of https://github.com/allmywant/mojito-js-delivery

allmywant committed:
Remove unused files

allmywant committed:
Remove the debug code

allmywant committed:
Merge branch 'mint-metrics:master' into master

allmywant committed:
add defensive check to prevent Mojito gets loading multi times

dapperdrop commented on gulpfile.js:
```diff
@@ -163,6 +163,31 @@ function build()
+        {
+            // add defensive check to prevent Mojito gets loading multi times
+            if (!config.allowMultiInstance) {
+                let content = fs.readFileSync(`dist/assets/js/${containerName}.js`, 'utf8');
```
Cheers @allmywant - this is a cool way of dealing with the issue! 

As there is some repetition in this block, would it be possible to refactor the code and wrap the guts of it in another function? E.g.
```
function insertDefensiveCheck(filename) {
    let content = fs.readFileSync(filename, 'utf8');
    let position = content.indexOf('Mojito =');
    if (position === -1) {
        position = content.indexOf('Mojito=');
    }
    if (position !== -1) {
        content = content.substr(0, position) + 'if (!window.Mojito || !Mojito.testObjects || !Object.keys(Mojito.testObjects).length) {' + content.substring(position) + '}';
        fs.writeFileSync(filename, content);
    }
}
```

```
// Inside the build function, replace the repetitive code block with:
if (!config.allowMultiInstance) {
    insertDefensiveCheck(`dist/assets/js/${containerName}.js`);
    insertDefensiveCheck(`dist/assets/js/${containerName}.pretty.js`);
}
```

allmywant replied:
@dapperdrop 
Good idea, I've wrapped the code into a `insertPreventDuplicateLoadingCheck` function.

allmywant committed:
Wrap the defensive check into a function

allmywant committed:
Add unit test for Prevent Mojito gets loading twice

kingo55 commented on gulpfile.js:
```diff
@@ -60,6 +60,17 @@ function test()
     return runTests();
 }
 
+function insertPreventDuplicateLoadingCheck(fileName) {
```
@dapperdrop @allmywant - minor issue, but I wonder if there's a clearer way to label this function. Like `insertDuplicateContainerCheck`.

allmywant replied:
@kingo55 
Renamed to `insertDuplicateContainerCheck`.

kingo55 commented on config.js:
```diff
@@ -0,0 +1,10 @@
+    s3BucketPRD: 'mojito-js/js',
+    // S3 Region
+    s3Region: 'ap-southeast-2'
+};
```
Do we need to show an example of `config.allowMultiInstance` setting in this file and in the example file too?

allmywant replied:
@kingo55 
added.

kingo55 commented on tests/test_suite.html:
```diff
@@ -927,6 +927,24 @@
+            po.src = '../dist/assets/js/mojito.js';
+            po.onload = function () {
+                window.setTimeout(function () {
+                    // if Mojit gets loading twice, then Object.keys(Mojito.testObjects).length should be 0.
```
`if Mojit gets loading twice` -> `if Mojito loads twice`

allmywant replied:
@kingo55 
Fixed.

kingo55 reviewed (commented):
@allmywant - just a few comments in here to address / discuss.

allmywant committed:
Minor updates

1. rename insertPreventDuplicateLoadingCheck to insertDuplicateContainerCheck
2. add an example of config.allowMultiInstance setting to config.js and config.example.js

kingo55 commented on tests/test_suite.html:
```diff
@@ -938,7 +938,7 @@
             po.onload = function () {
                 window.setTimeout(function () {
-                    // if Mojit gets loading twice, then Object.keys(Mojito.testObjects).length should be 0.
+                    // iif Mojito loads twice, then Object.keys(Mojito.testObjects).length should be 0.
```
@allmywant - sorry to be pedantic but it looks like there's an extra `i` in here now.

allmywant replied:
@kingo55 
Fixed :)

allmywant committed:
Fix typo
